experimental/compactdecimal/src/format.rs

```diff
@@ -0,0 +1,74 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::borrow::Cow;
+use fixed_decimal::{CompactDecimal, FixedDecimal};
+use writeable::Writeable;
+use zerovec::maps::ZeroMap2dCursor;
+
+use crate::compactdecimal::CompactDecimalFormatter;
+use crate::provider::{Count, PatternULE};
+
+/// An intermediate structure returned by [`CompactDecimalFormatter`](crate::CompactDecimalFormatter).
+/// Use [`Writeable`][Writeable] to render the formatted decimal to a string or buffer.
+pub struct FormattedCompactDecimal<'l> {
+    pub(crate) formatter: &'l CompactDecimalFormatter,
+    pub(crate) value: Cow<'l, CompactDecimal>,
+    pub(crate) plural_map: Option<ZeroMap2dCursor<'l, 'l, i8, Count, PatternULE>>,
```

Comment:
I had the same thought. There's not much use pre-resolving it since FormattedCompactDecimal is short-lived and generally intended to only be converted to a string or to parts; in fact, CompactDecimalFormatter::format is already doing more work than either FixedDecimalFormatter or DateTimeFormatter. Most of the work should live in FormattedCompactDecimal.

Reply:
> CompactDecimalFormatter::format is already doing more work than either FixedDecimalFormatter or DateTimeFormatter.

The reason for that is that the error checking of the fallible one (`format_compact_decimal`) needs to happen early enough that we can return a useful error (rather than `core::fmt::Error`).
That work being done, `FormattedCompactDecimal` naturally takes the shape it has.